### PR TITLE
Normalize include and exclude specs before file walking

### LINKED ISSUE #1874: Inconsistent slashes when creating an asp.net core project
After creating an ASP.NET Core project I added a target to print out the Content items. After executing that target I get the following results.

```
  Content: wwwroot/css\site.css
wwwroot/css\site.min.css
wwwroot/favicon.ico
wwwroot/file.ts
wwwroot/images\banner1.svg
wwwroot/images\banner2.svg
wwwroot/images\banner3.svg
wwwroot/images\banner4.svg
wwwroot/js\site.js
wwwroot/js\site.min.js
wwwroot/lib\bootstrap\.bower.json
wwwroot/lib\bootstrap\dist\css\bootstrap-theme.css
wwwroot/lib\bootstrap\dist\css\bootstrap-theme.css.map
wwwroot/lib\bootstrap\dist\css\bootstrap-theme.min.css
wwwroot/lib\bootstrap\dist\css\bootstrap-theme.min.css.map
wwwroot/lib\bootstrap\dist\css\bootstrap.css
wwwroot/lib\bootstrap\dist\css\bootstrap.css.map
wwwroot/lib\bootstrap\dist\css\bootstrap.min.css
wwwroot/lib\bootstrap\dist\css\bootstrap.min.css.map
wwwroot/lib\bootstrap\dist\fonts\glyphicons-halflings-regular.eot
wwwroot/lib\bootstrap\dist\fonts\glyphicons-halflings-regular.svg
wwwroot/lib\bootstrap\dist\fonts\glyphicons-halflings-regular.ttf
wwwroot/lib\bootstrap\dist\fonts\glyphicons-halflings-regular.woff
wwwroot/lib\bootstrap\dist\fonts\glyphicons-halflings-regular.woff2
wwwroot/lib\bootstrap\dist\js\bootstrap.js
wwwroot/lib\bootstrap\dist\js\bootstrap.min.js
wwwroot/lib\bootstrap\dist\js\npm.js
wwwroot/lib\bootstrap\LICENSE
wwwroot/lib\jquery-validation-unobtrusive\.bower.json
wwwroot/lib\jquery-validation-unobtrusive\jquery.validate.unobtrusive.js
wwwroot/lib\jquery-validation-unobtrusive\jquery.validate.unobtrusive.min.js
wwwroot/lib\jquery-validation\.bower.json
wwwroot/lib\jquery-validation\dist\additional-methods.js
wwwroot/lib\jquery-validation\dist\additional-methods.min.js
wwwroot/lib\jquery-validation\dist\jquery.validate.js
wwwroot/lib\jquery-validation\dist\jquery.validate.min.js
wwwroot/lib\jquery-validation\LICENSE.md
wwwroot/lib\jquery\.bower.json
wwwroot/lib\jquery\dist\jquery.js
wwwroot/lib\jquery\dist\jquery.min.js
wwwroot/lib\jquery\dist\jquery.min.map
wwwroot/lib\jquery\LICENSE.txt
wwwroot/test1.ts
...
```

The slashes here are not consistent. This is not a big deal because MSBuild will normalize these internally but if I have a custom task then the slashes not being consistent can cause issues I believe. I'd prefer these slashes to be consistent.

I think this is being caused by the globbing patterns for including wwwroot files.

### PR DESCRIPTION
**Pull request by cdmihai**
Avoids expensive string normalization on each discovered directory.

Improves performance (and probably memory) by not calling path aware comparisons and string normalizations during the recursive file walk. Instead, the user specs are normalized before the walk begins.

Also closes #1874 

Results based on building NHibernate.Test.csproj referenced in https://github.com/Microsoft/msbuild/issues/2392#issuecomment-330029024

image

### CODE REVIEW

**Comment by cdmihai** [comment]
@maca88

This should make this step unnecessary: https://github.com/maca88/msbuild/commit/9ca879af882a763d42d1bf109b2855670d839977#diff-64d47c5cbb59959b7b2fa6963672d596R776

**Comment by maca88** [comment]
Well done!

Not really related to this commit, but another optimization would be to get the subdirectories only once. Currently the subdirectories [are filled](https://github.com/Microsoft/msbuild/blob/master/src/Shared/FileMatcher.cs#L925) for the [current path](https://github.com/Microsoft/msbuild/blob/master/src/Shared/FileMatcher.cs#L717) and for each [exclude separately](https://github.com/Microsoft/msbuild/blob/master/src/Shared/FileMatcher.cs#L729). When we have excludes with a recursive pattern (`**`), which is always in case of [NHibernate](https://github.com/maca88/msbuild/commit/24ccceeafb1d7969107fc02c272bcd29909eea4d#diff-7a21f5e73644f08fe8b0dc9763a632aaR47) the `Directory.GetDirectories` is called for each folder multiplied by the number of excludes with the same search pattern (`null`). By calling it only once per directory, the cache per call would become useless (at least for this case) as we would never hit it. After I get home, I will try your changes to see their benefits on my computer as you seem to have a much faster machine :)

**Comment by cdmihai** [comment]
That's a neat observation! I've done a postprocess of a csproj using `Microsoft.NET.Sdk.Web` and all excludes have the recursive path wildcard in them.

(copied the exclude elements only)
```xml
        <DefaultItemExcludes>$(DefaultItemExcludes);**\node_modules\**;node_modules\**</DefaultItemExcludes>
        <DefaultItemExcludes>$(DefaultItemExcludes);**\jspm_packages\**;jspm_packages\**</DefaultItemExcludes>
        <DefaultItemExcludes>$(DefaultItemExcludes);**\bower_components\**;bower_components\**</DefaultItemExcludes>
        <DefaultItemExcludes>$(DefaultItemExcludes);**/*.user</DefaultItemExcludes>
        <DefaultItemExcludes>$(DefaultItemExcludes);**/*.*proj</DefaultItemExcludes>
        <DefaultItemExcludes>$(DefaultItemExcludes);**/*.sln</DefaultItemExcludes>
        <DefaultItemExcludes>$(DefaultItemExcludes);**/*.vssscc</DefaultItemExcludes>
        <DefaultItemExcludes>$(DefaultItemExcludes);packages/**</DefaultItemExcludes>
        <DefaultExcludesInProjectFolder>$(DefaultItemExcludesInProjectFolder);**/.*/**</DefaultExcludesInProjectFolder>
```

I'll comment in #2392 on sending some of your optimizations as PRs

**Comment by cdmihai** [comment]
@dotnet-bot test OSX Build for CoreCLR please

**Comment by AndyGerlicher** [comment]
Test | Type | Overall | Significant delta?  | Value
:--- | :---| :--- | :--- | :---
DotnetConsoleProject | Evaluation: Time (ms) | :red_circle: | yes | 27.0596 -> 27.1686 (0.403%)
DotnetConsoleProject | Evaluation: Memory | :ok_hand: | no | 4718123 -> 4719430 (0.028%)
DotnetWebProject | Evaluation: Time (ms) | :white_check_mark: | yes | 41.2793 -> 40.9544 (-0.787%)
DotnetWebProject | Evaluation: Memory | :ok_hand: | no | 6180417 -> 6182153 (0.028%)
DotnetMvcProject | Evaluation: Time (ms) | :red_circle: | yes | 79.2704 -> 79.7572 (0.614%)
DotnetMvcProject | Evaluation: Memory | :ok_hand: | no | 8451352 -> 8451429 (0.001%)
Picasso | Evaluation: Time (ms) | :ok_hand: | no | 223.2994 -> 223.7041 (0.181%)
Picasso | Evaluation: Memory | ::ok_hand: | no | 32038700 -> 32037420 (-0.004%)
SmallP2POldCsproj | Evaluation: Time (ms) | :white_check_mark: | yes | 36.1425 -> 36.0586 (-0.232%)
SmallP2POldCsproj | Evaluation: Memory | :ok_hand: | no | 5173779 -> 5174157 (0.007%)
Generated_100_100_v150 | Evaluation: Time (ms) | :white_check_mark: | yes | 1384.9986 -> 1379.1015 (-0.426%)
Generated_100_100_v150 | Evaluation: Memory | ::ok_hand: | no | 174508124 -> 174507796 (0%)
LargeP2POldCsproj | Evaluation: Time (ms) | :white_check_mark: | yes | 639.4464 -> 633.5188 (-0.927%)
LargeP2POldCsproj | Evaluation: Memory | ::ok_hand: | no | 68544245 -> 68541894 (-0.003%)
roslyn | Evaluation: Time (ms) | :red_circle: | yes | 5178.9291 -> 5233.5785 (1.055%)
roslyn | Evaluation: Memory | :white_check_mark: | yes | 579198997 -> 572198814 (-1.209%)

File: src/Shared/FileMatcher.cs [hunk 1]
```diff
@@ -780,7 +780,7 @@ private static void GetFilesRecursive
                         for (int i = 0; i < excludeNextSteps.Length; i++)
                         {
                             if (excludeNextSteps[i].Subdirs != null &&
-                                excludeNextSteps[i].Subdirs.Any(excludedDir => FileUtilities.PathsEqual(excludedDir, subdir)))
+                                excludeNextSteps[i].Subdirs.Any(excludedDir => excludedDir == subdir))
```
Comment (AndyGerlicher) [inline src/Shared/FileMatcher.cs]: Do we need to avoid linq here? How hot is this path?

File: src/Shared/FileMatcher.cs [hunk 2]
```diff
@@ -1582,12 +1579,108 @@ enum SearchAction
...
             result.SearchData = searchData;
-            result.BaseDirectory = fixedDirectoryPart;
-            result.RemainingWildcardDirectory = wildcardDirectoryPart;
+            result.BaseDirectory = Normalize(fixedDirectoryPart);
+            result.RemainingWildcardDirectory = Normalize(wildcardDirectoryPart);
 
             return SearchAction.RunSearch;
         }
 
+        // Replace all slashes to the OS slash, collapse multiple slashes into one, trim trailing slashes
```
Comment (AndyGerlicher) [inline src/Shared/FileMatcher.cs]: Can you make this xmldoc comment please?

File: src/Shared/FileMatcher.cs (hunk 1 above)
Comment (cdmihai) [reply src/Shared/FileMatcher.cs]: Pretty hot. I'll go ahead and remove all Linq operations from this method

**Comment by AndyGerlicher** [comment]
@dotnet-bot test Windows_NT Build for Full please

**Comment by AndyGerlicher** [comment]
@dotnet-bot test Windows_NT Build for Full please

**Comment by cdmihai** [comment]
@dotnet-bot test Windows_NT Build for CoreCLR please

File: src/Shared/FileMatcher.cs (hunk 1 above)
Comment (cdmihai) [reply src/Shared/FileMatcher.cs]: On the other hand, @maca88's changes remove the need for this linq operation, so best to leave like this to reduce the future merge conflicts :)

**Pull request merged**